zeebe/exporters/camunda-exporter/src/test/java/io/camunda/exporter/utils/ContainerizedSearchDatabaseCallbackDelegate.java

```diff
@@ -0,0 +1,104 @@
+
+import static io.camunda.exporter.config.ConnectionTypes.ELASTICSEARCH;
+import static io.camunda.exporter.config.ConnectionTypes.OPENSEARCH;
+
+import co.elastic.clients.elasticsearch.ElasticsearchClient;
+import io.camunda.exporter.config.ConnectionTypes;
+import io.camunda.exporter.config.ExporterConfiguration;
+import io.camunda.search.connect.es.ElasticsearchConnector;
+import io.camunda.search.connect.os.OpensearchConnector;
+import io.camunda.zeebe.test.util.testcontainers.TestSearchContainers;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.opensearch.client.opensearch.OpenSearchClient;
+import org.opensearch.testcontainers.OpensearchContainer;
+import org.testcontainers.elasticsearch.ElasticsearchContainer;
+
+public class ContainerizedSearchDatabaseCallbackDelegate implements SearchDatabaseCallbackDelegate {
+
+  protected SearchClientAdapter elsClientAdapter;
+  protected SearchClientAdapter osClientAdapter;
+
+  private ElasticsearchContainer elsContainer;
+  private OpensearchContainer<?> osContainer;
+
+  private ElasticsearchClient elsClient;
+  private OpenSearchClient osClient;
+
+  private final List<AutoCloseable> closeables = new ArrayList<>();
+
+  @Override
+  public void afterAll(final ExtensionContext context) throws IOException {
+    closeables.parallelStream()
+        .forEach(
+            c -> {
+              try {
+                c.close();
+              } catch (final Exception e) {
+                throw new RuntimeException(e);
+              }
+            });
+  }
+
+  @Override
+  public void afterEach(final ExtensionContext context) throws IOException {
+    if (context.getDisplayName().equals(ConnectionTypes.ELASTICSEARCH.getType())) {
+      elsClient.indices().delete(req -> req.index("*"));
+      elsClient.indices().deleteIndexTemplate(req -> req.name("*"));
+    } else if (context.getDisplayName().equals(ConnectionTypes.OPENSEARCH.getType())) {
+      osClient.indices().delete(req -> req.index("*"));
+      osClient.indices().deleteIndexTemplate(req -> req.name("*"));
+    }
+  }
+
+  @Override
+  public void beforeAll(final ExtensionContext context) throws IOException {
+    elsContainer = TestSearchContainers.createDefeaultElasticsearchContainer();
+    osContainer = TestSearchContainers.createDefaultOpensearchContainer();
+
+    elsContainer.start();
+    osContainer.start();
+
+    final var osConfig = getConfigWithConnectionDetails(OPENSEARCH);
+    osClient = new OpensearchConnector(osConfig.getConnect()).createClient();
+    osClientAdapter = new SearchClientAdapter(osClient);
+
+    final var elsConfig = getConfigWithConnectionDetails(ELASTICSEARCH);
+    elsClient = new ElasticsearchConnector(elsConfig.getConnect()).createClient();
+    elsClientAdapter = new SearchClientAdapter(elsClient);
+
+    closeables.add(elsContainer);
+    closeables.add(osContainer);
+  }
```

Comment:
I think an alternative with less changes would be:

```suggestion
    
    // pre-create os client config
    final var osConfig = getDefaultConfig(OPENSEARCH);
 
    final var shouldRunAgainstAWSOS =
        Boolean.parseBoolean(            Optional.ofNullable(System.getProperty(SHOULD_RUN_AGAINST_AWS)).orElse("false"));
    if (shouldRunAgainstAWSOS) {
       // test should run against AWS Opensearch (external dependency)
       final var openSearchAwsInstanceUrl =
       
 Optional.ofNullable(System.getProperty(AWS_OPENSEARCH_SYSTEM_PROPERTY_NAME)).orElseThrow();
      osConfig.getConnect().setUrl(openSearchAwsInstanceUrl);
    } else {
        elsContainer = TestSearchContainers.createDefeaultElasticsearchContainer();
        osContainer = TestSearchContainers.createDefaultOpensearchContainer();
    
        elsContainer.start();
        osContainer.start();
   
        osConfig.getConnect().setUrl(osContainer.getHttpHostAddress());
        
        final var elsConfig = getDefaultConfig(ELASTICSEARCH);
        elsConfig.getConnect().setUrl(elsContainer.getHttpHostAddress());
        elsClient = new ElasticsearchConnector(elsConfig.getConnect()).createClient();
        elsClientAdapter = new SearchClientAdapter(elsClient);
        
        closeables.add(elsContainer);
        closeables.add(osContainer);
    }
    
    osClient = new OpensearchConnector(osConfig.getConnect()).createClient();
    osClientAdapter = new SearchClientAdapter(osClient);
  }
  
  
    @Override
  public ExporterConfiguration getDefaultConfig(
      final ConnectionTypes connectionType) {
    final var config = new ExporterConfiguration();
    config.getIndex().setPrefix(CamundaExporterITInvocationProvider.CONFIG_PREFIX);
    config.getBulk().setSize(1); // force flushing on the first record
    config.getConnect().setType(connectionType.getType());
    return config;
  }
  
```

Reply:
Good suggestion. I was worried that if I leave AWS OS in previous implementation (`CamundaExporterITInvocationProvider`), we might still at some point have additional logic specifically related to AWS OS or docker containers where at which moment it may become painful to refactor; on the other hand we don't have this issue now so maybe it is not a bad idea at all.
I will rewrite as you suggest.